### ISSUE TITLE
[Material3] DropdownMenu Text Overflow

### ISSUE THREAD

**Issue by recjason** [flutter/flutter#126882]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Use dartpad here: https://api.flutter.dev/flutter/material/DropdownMenu-class.html
2. Update to multiply the string (create a long string so text overflows): 

```dart
    for (final ColorLabel color in ColorLabel.values) {
      colorEntries.add(
        DropdownMenuEntry<ColorLabel>(
            value: color, label: color.label * 300, enabled: color.label != 'Grey'),
      );
    }
```

3.  Update `DropdownMenu` definition to pass a `width`. (This stops the dropdown selector from overflowing)
4.  Update `DropdownMenu` definition to pass a MenuStyle with `maximumWidth`. (This stops the dropdown selector from overflowing)
5. Neither changing `DropdownMenu.textStyle` or `DropdownMenuEntry.style` (`ButtonStyle.textStyle`) allows changing how the text overflows.

### Expected results

Can choose whether the text is clipped/ellipsis/wraps etc.

### Actual results

Text Render overflows

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [DropdownMenu]s. The first dropdown menu has an outlined border.

void main() => runApp(const DropdownMenuExample());

class DropdownMenuExample extends StatefulWidget {
  const DropdownMenuExample({super.key});

  @override
  State<DropdownMenuExample> createState() => _DropdownMenuExampleState();
}

class _DropdownMenuExampleState extends State<DropdownMenuExample> {
  final TextEditingController colorController = TextEditingController();
  final TextEditingController iconController = TextEditingController();
  ColorLabel? selectedColor;
  IconLabel? selectedIcon;

  @override
  Widget build(BuildContext context) {
    final List<DropdownMenuEntry<ColorLabel>> colorEntries =
        <DropdownMenuEntry<ColorLabel>>[];
    for (final ColorLabel color in ColorLabel.values) {
      colorEntries.add(
        DropdownMenuEntry<ColorLabel>(
            value: color, label: color.label * 300, enabled: color.label != 'Grey' ),
      );
    }

    final List<DropdownMenuEntry<IconLabel>> iconEntries =
        <DropdownMenuEntry<IconLabel>>[];
    for (final IconLabel icon in IconLabel.values) {
      iconEntries
          .add(DropdownMenuEntry<IconLabel>(value: icon, label: icon.label));
    }

    return MaterialApp(
      theme: ThemeData(
        useMaterial3: true,
        colorSchemeSeed: Colors.green,
      ),
      home: Scaffold(
        body: SafeArea(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.symmetric(vertical: 20),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    DropdownMenu<ColorLabel>(
                      controller: colorController,
                      label: const Text('Color'),
                      dropdownMenuEntries: colorEntries,
                      onSelected: (ColorLabel? color) {
                        setState(() {
                          selectedColor = color;
                        });
                      },
                      width: 200,
                      textStyle: TextStyle(overflow: TextOverflow.ellipsis),
                      menuStyle: MenuStyle(maximumSize: MaterialStatePropertyAll(Size.fromWidth(200))),
                    ),
                    const SizedBox(width: 20),
                    DropdownMenu<IconLabel>(
                      controller: iconController,
                      enableFilter: true,
                      leadingIcon: const Icon(Icons.search),
                      label: const Text('Icon'),
                      dropdownMenuEntries: iconEntries,
                      inputDecorationTheme: const InputDecorationTheme(
                        filled: true,
                        contentPadding: EdgeInsets.symmetric(vertical: 5.0),
                      ),
                      onSelected: (IconLabel? icon) {
                        setState(() {
                          selectedIcon = icon;
                        });
                      },
                    )
                  ],
                ),
              ),
              if (selectedColor != null && selectedIcon != null)
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Text(
                        'You selected a ${selectedColor?.label} ${selectedIcon?.label}'),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 5),
                      child: Icon(
                        selectedIcon?.icon,
                        color: selectedColor?.color,
                      ),
                    )
                  ],
                )
              else
                const Text('Please select a color and an icon.')
            ],
          ),
        ),
      ),
    );
  }
}

enum ColorLabel {
  blue('Blue', Colors.blue),
  pink('Pink', Colors.pink),
  green('Green', Colors.green),
  yellow('Yellow', Colors.yellow),
  grey('Grey', Colors.grey);

  const ColorLabel(this.label, this.color);
  final String label;
  final Color color;
}

enum IconLabel {
  smile('Smile', Icons.sentiment_satisfied_outlined),
  cloud(
    'Cloud',
    Icons.cloud_outlined,
  ),
  brush('Brush', Icons.brush_outlined),
  heart('Heart', Icons.favorite);

  const IconLabel(this.label, this.icon);
  final String label;
  final IconData icon;
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
from api.flutter.dev ....
```

</details>

**Comment by HansMuller** [comment]
Here's a simpler code sample that demonstrates the problem: 

```dart
import 'package:flutter/material.dart';

enum ColorItem {
  blue('Blue', Colors.blue),
  pink('Pink', Colors.pink),
  green('Green', Colors.green),
  yellow('Yellow', Colors.yellow),
  grey('Grey', Colors.grey);

  const ColorItem(this.label, this.color);
  final String label;
  final Color color;
}

class DropdownMenuIssue extends StatefulWidget {
  const DropdownMenuIssue({super.key});

  @override
  State<DropdownMenuIssue> createState() => _DropdownMenuIssueState();
}

class _DropdownMenuIssueState extends State<DropdownMenuIssue> {
  late final TextEditingController controller;

  @override
  void initState() {
    super.initState();
    controller = TextEditingController();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    return Scaffold(
      body: Center(
        child: DropdownMenu<ColorItem>(
          width: 200,
          controller: controller,
          initialSelection: ColorItem.green,
          label: const Text('Color'),
          onSelected: (ColorItem? color) {
            print('Selected $color');
          },
          dropdownMenuEntries: ColorItem.values.map<DropdownMenuEntry<ColorItem>>((ColorItem item) {
            return DropdownMenuEntry<ColorItem>(
              value: item,
              label: item.label * 20,
            );
          }).toList(),
        ),
      ),
    );
  }
}

void main() {
  runApp(
    MaterialApp(home: const DropdownMenuIssue()),
  );
}
```

**Comment by HansMuller** [comment]
If the maxWidth of the Text widget created for each dropdown menu item's label is constrained when DropdownMenu.width is specified, then the menu items will wrap, instead of overflowing:

![Screenshot 2023-08-21 at 6 03 59 PM](https://github.com/flutter/flutter/assets/1377460/c3f63373-58eb-41dc-b62e-8906f7c388b8)

For the app to have additional control over how the label is rendered, it might be best to allow a DropdownMenuEntry `labelWidget` to be specified, along with the label. That would make it possible to configure the number of lines the text widget can expand to, and how overflow is handled. And if you wanted to use something other than `Text(label)` you could do that too.

Sound reasonable?

**Comment by feinstein** [comment]
@HansMuller I am facing this issue and yes, it sounds reasonable to me.

**Comment by HansMuller** [comment]
The PR linked above fixes the problem reported here and enables creating your own label widget with the new `DropdownMenuEntry` ` labelWidget` property

**Comment by feinstein** [comment]
I tried to solve this yesterday, got a solution very close to yours, but I found 2 weird behaviours that I wanted to investigate further.

First, the Flutter Inspector overlay wasn't highlighting the menu items, and clicking for finding the widget was closing the drop down, instead of ignoring the click.

Second, there are 2 Rows inside the MenuItemLabel, and they were messing with the constraints.

I tried constraining the label, but I didn't add your padding, maybe that was what was missing. This fixed partially, as the label wasn't precisely the width of the DropDownMenu. Then I tried to constrain the MenuButton, but that didn't work either, the width was correct, but the 2 Rows made the text overflow.

I also used maxWidth at first, but then I thought this should be a tight constraint, to match the DropDownMenu width, and this made it worse.

I will test your PR and try to understand the difference that padding makes. Thanks Hans.
